### ISSUE TITLE
[bug] WIn10/MacOS tauri 1.2. Asset protocol stopped work.

### ISSUE THREAD

**Issue by Bukmopbl4** [tauri-apps/tauri#5582]
### Describe the bug

In version 1.1, I loaded resources via api window.__TAURI__.path.resolveResource. All permissions were set and everything worked perfectly. After updating to v1.2 I get 404 in all requests for asset:// protocol.

### Platform and versions

Environment
  › OS: Mac OS 13.0.0 X64
  › Node.js: 16.17.1
  › npm: 8.15.0
  › pnpm: Not installed!
  › yarn: Not installed!
  › rustup: 1.25.1
  › rustc: 1.64.0
  › cargo: 1.64.0
  › Rust toolchain: stable-x86_64-apple-darwin 

Packages
WARNING: no lock files found, defaulting to npm
  › @tauri-apps/cli [NPM]: 1.2.0
  › @tauri-apps/api [NPM]: Not installed!
  › tauri [RUST]: 1.2.0,
  › tauri-build [RUST]: 1.2.0,
  › tao [RUST]: 0.15.0,
  › wry [RUST]: 0.22.0,

App
  › build-type: bundle
  › CSP: unset
  › distDir: ../src
  › devPath: ../src
package.json not found

App directory structure
  ├─ src-tauri
  ├─ .git
  ├─ .vscode
  └─ src

**Comment by Bukmopbl4** [comment]
when you roll back to version 1.1, the problem goes away.

**Comment by FabianLars** [comment]
~~Linux and Windows seem to work fine.~~ nevermind, didn't use resolveResource but hard-coded values. I can't re-test it today.

**Comment by Bukmopbl4** [comment]
> Linux and Windows seem to work fine.

let me try again on windows 10, but on macOS ventura I checked it many times and the assets do not work.

**Comment by FabianLars** [comment]
~~I believe you, especially since someone on discord has the same issue on macos. My comment was more for the team so that they focus on the macos changes~~ see above, also i didn't see that the title contained `Win10`

**Comment by Bukmopbl4** [comment]
on win10 resolveResource return something like this `\\?\C:\Users\...`. What is this `\\?\ `? If I delete from returned path this line `\\?\` my window can load the asset on win10.

**Comment by lucasfernog** [comment]
The problem is that we're canonicalizing the resource path now. The asset protocol needs to handle that. I'll check this later.

**Comment by lucasfernog** [comment]
@Bukmopbl4 how are you sending the requests? I tried this:

```javascript
let a = await window.__TAURI__.path.resolveResource('a.txt')
let p = window.__TAURI__.tauri.convertFileSrc(a)
fetch(p)
```

and it works just fine on Windows (with `["$RESOURCE/**"]` in the scope). Make sure you're not interpolating the path with `asset://` directly, the convertFileSrc is the preferred way to do this. If you're doing it manually, change `asset://` with `asset://localhost` since we're now validating the custom protocol request URI.

**Comment by Bukmopbl4** [comment]
> how are you sending the requests? I tried this:

i will try with convertFileSrc but normally adding asset:// to getted path worked fine on MacOS (obviously with ["$RESOURCE/**"] in the scope)

**Comment by lucasfernog** [comment]
Yeah the path is now validated so `asset://$PATH` won't work (and also isn't great for cross platform applications, that's why we provide the convertFileSrc function).

**Comment by Bukmopbl4** [comment]
> @Bukmopbl4 how are you sending the requests? I tried this:
> 
> ```js
> let a = await window.__TAURI__.path.resolveResource('a.txt')
> let p = window.__TAURI__.tauri.convertFileSrc(a)
> fetch(p)
> ```
> 
> and it works just fine on Windows (with `["$RESOURCE/**"]` in the scope). Make sure you're not interpolating the path with `asset://` directly, the convertFileSrc is the preferred way to do this. If you're doing it manually, change `asset://` with `asset://localhost` since we're now validating the custom protocol request URI.

Works great with convertFileSrc
